File: shared/src/main/java/com/runtracker/shared/data/repository/GamificationRepository.kt
package com.runtracker.shared.data.repository

import com.runtracker.shared.data.db.GamificationDao
import com.runtracker.shared.data.model.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import java.util.*

class GamificationRepository(
    private val gamificationDao: GamificationDao
) {
    // User Gamification
    fun getUserGamification(): Flow<UserGamification?> = gamificationDao.getUserGamification()
    
    suspend fun getOrCreateUserGamification(): UserGamification {
        val existing = gamificationDao.getUserGamificationOnce()
        if (existing != null) return existing
        
        val newGamification = UserGamification()
        gamificationDao.insertOrUpdateGamification(newGamification)
        return newGamification
    }
    
    suspend fun addXp(amount: Int, reason: XpReason, relatedId: Long? = null) {
        val current = getOrCreateUserGamification()
        val newTotalXp = current.totalXp + amount
        val newLevel = UserGamification.levelForXp(newTotalXp)
        
        gamificationDao.addXp(amount.toLong(), newLevel)
        gamificationDao.insertXpTransaction(XpTransaction(
            amount = amount,
            reason = reason,
            relatedId = relatedId
        ))
        
        // Check for level up bonus
        if (newLevel > current.currentLevel) {
            val levelUpBonus = newLevel * 25
            gamificationDao.addXp(levelUpBonus.toLong(), newLevel)
            gamificationDao.insertXpTransaction(XpTransaction(
                amount = levelUpBonus,
                reason = XpReason.LEVEL_UP_BONUS
            ))
        }
    }
    
    suspend fun recordWorkoutCompleted(distanceMeters: Double, durationMinutes: Int, runId: Long) {
        val gamification = getOrCreateUserGamification()
        
        // Update streak
        val today = getStartOfDay(System.currentTimeMillis())
        val lastActivity = gamification.lastActivityDate?.let { getStartOfDay(it) }
        val yesterday = getStartOfDay(System.currentTimeMillis() - 24 * 60 * 60 * 1000)
        
        val newStreak = when {
            lastActivity == today -> gamification.currentStreak // Already worked out today
            lastActivity == yesterday -> gamification.currentStreak + 1 // Continuing streak
            lastActivity == null -> 1 // First workout
            else -> 1 // Streak broken, start new
        }
        
        val newLongest = maxOf(newStreak, gamification.longestStreak)
        gamificationDao.updateStreak(newStreak, newLongest, System.currentTimeMillis())
        
        // Update workout stats
        gamificationDao.incrementWorkoutStats(distanceMeters, durationMinutes)
        
        // Calculate and add XP
        val xp = XpRewards.calculateWorkoutXp(distanceMeters, durationMinutes, newStreak)
        addXp(xp, XpReason.WORKOUT_COMPLETED, runId)
        
        // Update daily rings
        updateDailyRings(durationMinutes, distanceMeters)
        
        // Check achievements
        checkAchievements()
    }
    
    suspend fun updateDailyRings(exerciseMinutes: Int, distanceMeters: Double) {
        val today = getStartOfDay(System.currentTimeMillis())
        var rings = gamificationDao.getDailyRingsOnce(today)
        
        if (rings == null) {
            rings = DailyRings(date = today)
            gamificationDao.insertOrUpdateDailyRings(rings)
        }
        
        gamificationDao.addActivityMinutes(today, exerciseMinutes, exerciseMinutes)
        gamificationDao.addDistance(today, distanceMeters)
        
        // Check if all rings closed
        val updatedRings = gamificationDao.getDailyRingsOnce(today)
        if (updatedRings != null) {
            val allClosed = updatedRings.moveProgress >= 1f && 
                           updatedRings.exerciseProgress >= 1f
            if (allClosed && !updatedRings.allRingsClosed) {
                gamificationDao.setAllRingsClosed(today, true)
                addXp(XpRewards.DAILY_GOAL_BONUS, XpReason.DAILY_GOAL_COMPLETED)
            }
        }
    }
    
    suspend fun checkAchievements() {
        val gamification = getOrCreateUserGamification()
        val achievements = gamificationDao.getAllAchievementsOnce()
        
        for (achievement in achievements) {
            val userAchievement = gamificationDao.getUserAchievement(achievement.id)
            if (userAchievement?.isUnlocked == true) continue
            
            val currentProgress = when (achievement.category) {
                AchievementCategory.DISTANCE -> (gamification.totalDistanceMeters / 1000).toInt()
                AchievementCategory.WORKOUTS -> gamification.totalWorkoutsCompleted
                AchievementCategory.STREAKS -> gamification.longestStreak
                AchievementCategory.CONSISTENCY -> gamification.weeklyWorkouts
                else -> userAchievement?.progress ?: 0
            }
            
            val isUnlocked = currentProgress >= achievement.requirement
            
            if (userAchievement == null) {
                gamificationDao.insertUserAchievement(UserAchievement(
                    achievementId = achievement.id,
                    progress = currentProgress,
                    isUnlocked = isUnlocked,
                    unlockedAt = if (isUnlocked) System.currentTimeMillis() else 0
                ))
            } else if (currentProgress != userAchievement.progress || isUnlocked != userAchievement.isUnlocked) {
                gamificationDao.updateAchievementProgress(
                    achievement.id,
                    currentProgress,
                    isUnlocked,
                    if (isUnlocked && !userAchievement.isUnlocked) System.currentTimeMillis() else userAchievement.unlockedAt
                )
            }
            
            // Award XP for newly unlocked achievement
            if (isUnlocked && userAchievement?.isUnlocked != true) {
                addXp(achievement.xpReward, XpReason.ACHIEVEMENT_UNLOCKED)
            }
        }
    }
    
    suspend fun recordPersonalBest(runId: Long) {
        addXp(XpRewards.PERSONAL_BEST, XpReason.PERSONAL_BEST, runId)
        
        // Update PB achievement progress
        val pbAchievements = listOf("pb_first", "pb_5", "pb_10")
        for (achievementId in pbAchievements) {
            val userAchievement = gamificationDao.getUserAchievement(achievementId)
            val newProgress = (userAchievement?.progress ?: 0) + 1
            val achievement = gamificationDao.getAchievementById(achievementId) ?: continue
            val isUnlocked = newProgress >= achievement.requirement
            
            gamificationDao.updateAchievementProgress(
                achievementId,
                newProgress,
                isUnlocked,
                if (isUnlocked) System.currentTimeMillis() else 0
            )
            
            if (isUnlocked && userAchievement?.isUnlocked != true) {
                addXp(achievement.xpReward, XpReason.ACHIEVEMENT_UNLOCKED)
            }
        }
    }
    
    // Achievements
    fun getAllAchievements(): Flow<List<Achievement>> = gamificationDao.getAllAchievements()
    fun getUserAchievements(): Flow<List<UserAchievement>> = gamificationDao.getAllUserAchievements()
    fun getUnlockedAchievements(): Flow<List<UserAchievement>> = gamificationDao.getUnlockedAchievements()
    
    suspend fun getUnshownAchievementNotifications(): List<UserAchievement> = 
        gamificationDao.getUnshownAchievementNotifications()
    
    suspend fun markAchievementNotificationShown(achievementId: String) {
        gamificationDao.markAchievementNotificationShown(achievementId)
    }
    
    // Daily Rings
    fun getDailyRings(date: Long): Flow<DailyRings?> = gamificationDao.getDailyRings(date)
    fun getRecentDailyRings(days: Int = 7): Flow<List<DailyRings>> = gamificationDao.getRecentDailyRings(days)
    
    suspend fun getTodayRings(): DailyRings {
        val today = getStartOfDay(System.currentTimeMillis())
        var rings = gamificationDao.getDailyRingsOnce(today)
        if (rings == null) {
            rings = DailyRings(date = today)
            gamificationDao.insertOrUpdateDailyRings(rings)
        }
        return rings
    }
    
    // XP Transactions
    fun getRecentXpTransactions(limit: Int = 20): Flow<List<XpTransaction>> = 
        gamificationDao.getRecentXpTransactions(limit)
    
    // Seed default achievements
    suspend fun seedDefaultAchievements() {
        val existing = gamificationDao.getAllAchievementsOnce()
        if (existing.isEmpty()) {
            gamificationDao.insertAchievements(DefaultAchievements.getAll())
        }
    }
    
    // Helper
    private fun getStartOfDay(timestamp: Long): Long {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = timestamp
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        return calendar.timeInMillis
    }
}

File: iosApp/GoSteady/Sources/Repositories/GamificationRepository.swift
import Foundation
import SwiftData

final class GamificationRepository {
    private let context: ModelContext

    init(context: ModelContext) {
        self.context = context
    }

    // MARK: - User Gamification

    func fetchUserGamification() -> UserGamification? {
        let descriptor = FetchDescriptor<UserGamification>()
        return try? context.fetch(descriptor).first
    }

    func getOrCreateUserGamification() -> UserGamification {
        if let existing = fetchUserGamification() { return existing }
        let gamification = UserGamification()
        context.insert(gamification)
        try? context.save()
        return gamification
    }

    // MARK: - XP

    func addXp(amount: Int, reason: XpReason, relatedId: UUID? = nil) {
        let current = getOrCreateUserGamification()
        current.totalXp += Int64(amount)
        let newLevel = UserGamification.levelForXp(current.totalXp)

        let leveledUp = newLevel > current.currentLevel
        current.currentLevel = newLevel
        try? context.save()

        let transaction = XpTransaction(amount: amount, reason: reason, relatedId: relatedId)
        context.insert(transaction)
        try? context.save()

        if leveledUp {
            let bonus = newLevel * 25
            current.totalXp += Int64(bonus)
            current.currentLevel = UserGamification.levelForXp(current.totalXp)
            let bonusTx = XpTransaction(amount: bonus, reason: .levelUpBonus)
            context.insert(bonusTx)
            try? context.save()
        }
    }

    // MARK: - Workout Recording

    func recordWorkoutCompleted(distanceMeters: Double, durationMinutes: Int, workoutId: UUID) {
        let gamification = getOrCreateUserGamification()

        // Update streak
        let today = Date().startOfDay
        let yesterday = Calendar.current.date(byAdding: .day, value: -1, to: today)!.startOfDay
        let lastActivity = gamification.lastActivityDate?.startOfDay

        let newStreak: Int
        if lastActivity == today {
            newStreak = gamification.currentStreak
        } else if lastActivity == yesterday {
            newStreak = gamification.currentStreak + 1
        } else {
            newStreak = 1
        }

        gamification.currentStreak = newStreak
        gamification.longestStreak = max(newStreak, gamification.longestStreak)
        gamification.lastActivityDate = Date()
        gamification.totalWorkoutsCompleted += 1
        gamification.totalDistanceMeters += distanceMeters
        gamification.totalDurationMinutes += durationMinutes
        try? context.save()

        let xp = XpRewards.calculateWorkoutXp(
            distanceMeters: distanceMeters,
            durationMinutes: durationMinutes,
            currentStreak: newStreak
        )
        addXp(amount: xp, reason: .workoutCompleted, relatedId: workoutId)

        updateDailyRings(exerciseMinutes: durationMinutes, distanceMeters: distanceMeters)
        checkAchievements()
    }

    // MARK: - Daily Rings

    func getTodayRings() -> DailyRings {
        let today = Date().startOfDay
        let descriptor = FetchDescriptor<DailyRings>(predicate: #Predicate { $0.date == today })
        if let existing = try? context.fetch(descriptor).first {
            return existing
        }
        let rings = DailyRings(date: today)
        context.insert(rings)
        try? context.save()
        return rings
    }

    func fetchDailyRings(date: Date) -> DailyRings? {
        let dayStart = date.startOfDay
        let descriptor = FetchDescriptor<DailyRings>(predicate: #Predicate { $0.date == dayStart })
        return try? context.fetch(descriptor).first
    }

    func fetchRecentDailyRings(days: Int = 7) -> [DailyRings] {
        let cutoff = Calendar.current.date(byAdding: .day, value: -days, to: Date())?.startOfDay ?? Date()
        let descriptor = FetchDescriptor<DailyRings>(
            predicate: #Predicate { $0.date >= cutoff },
            sortBy: [SortDescriptor(\.date, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func updateDailyRings(exerciseMinutes: Int, distanceMeters: Double) {
        let rings = getTodayRings()
        rings.moveMinutes += exerciseMinutes
        rings.exerciseMinutes += exerciseMinutes
        rings.distanceMeters += distanceMeters

        let allClosed = rings.moveProgress >= 1.0 && rings.exerciseProgress >= 1.0
        if allClosed && !rings.allRingsClosed {
            rings.allRingsClosed = true
            addXp(amount: XpRewards.dailyGoalBonus, reason: .dailyGoalCompleted)
        }
        try? context.save()
    }

    // MARK: - Achievements

    func fetchAllAchievements() -> [Achievement] {
        let descriptor = FetchDescriptor<Achievement>(sortBy: [SortDescriptor(\.name)])
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchUserAchievements() -> [UserAchievement] {
        let descriptor = FetchDescriptor<UserAchievement>(sortBy: [SortDescriptor(\.unlockedAt, order: .reverse)])
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchUnlockedAchievements() -> [UserAchievement] {
        let descriptor = FetchDescriptor<UserAchievement>(
            predicate: #Predicate { $0.isUnlocked == true },
            sortBy: [SortDescriptor(\.unlockedAt, order: .reverse)]
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchUnshownNotifications() -> [UserAchievement] {
        let descriptor = FetchDescriptor<UserAchievement>(
            predicate: #Predicate { $0.isUnlocked == true && $0.notificationShown == false }
        )
        return (try? context.fetch(descriptor)) ?? []
    }

    func markNotificationShown(achievementId: String) {
        let descriptor = FetchDescriptor<UserAchievement>(
            predicate: #Predicate { $0.achievementId == achievementId }
        )
        if let ua = try? context.fetch(descriptor).first {
            ua.notificationShown = true
            try? context.save()
        }
    }

    func checkAchievements() {
        let gamification = getOrCreateUserGamification()
        let achievements = fetchAllAchievements()

        for achievement in achievements {
            let achievementId = achievement.id
            let descriptor = FetchDescriptor<UserAchievement>(
                predicate: #Predicate { $0.achievementId == achievementId }
            )
            let userAchievement = try? context.fetch(descriptor).first

            if userAchievement?.isUnlocked == true { continue }

            let currentProgress: Int
            switch achievement.category {
            case .distance:    currentProgress = Int(gamification.totalDistanceMeters / 1000)
            case .workouts:    currentProgress = gamification.totalWorkoutsCompleted
            case .streaks:     currentProgress = gamification.longestStreak
            case .consistency: currentProgress = gamification.weeklyWorkouts
            default:           currentProgress = userAchievement?.progress ?? 0
            }

            let isUnlocked = currentProgress >= achievement.requirement

            if let ua = userAchievement {
                if currentProgress != ua.progress || isUnlocked != ua.isUnlocked {
                    ua.progress = currentProgress
                    if isUnlocked && !ua.isUnlocked {
                        ua.isUnlocked = true
                        ua.unlockedAt = Date()
                        addXp(amount: achievement.xpReward, reason: .achievementUnlocked)
                    }
                }
            } else {
                let newUA = UserAchievement(
                    achievementId: achievement.id,
                    unlockedAt: isUnlocked ? Date() : Date.distantPast,
                    progress: currentProgress,
                    isUnlocked: isUnlocked
                )
                context.insert(newUA)
                if isUnlocked {
                    addXp(amount: achievement.xpReward, reason: .achievementUnlocked)
                }
            }
        }
        try? context.save()
    }

    func recordPersonalBest(workoutId: UUID) {
        addXp(amount: XpRewards.personalBest, reason: .personalBest, relatedId: workoutId)
    }

    // MARK: - XP Transactions

    func fetchRecentXpTransactions(limit: Int = 20) -> [XpTransaction] {
        var descriptor = FetchDescriptor<XpTransaction>(sortBy: [SortDescriptor(\.timestamp, order: .reverse)])
        descriptor.fetchLimit = limit
        return (try? context.fetch(descriptor)) ?? []
    }

    // MARK: - Seed

    func seedDefaultAchievements() {
        let existing = fetchAllAchievements()
        guard existing.isEmpty else { return }
        for a in DefaultAchievements.all {
            context.insert(a)
        }
        try? context.save()
    }
}
